documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+                           IEquatable<Vector2>,
+                           IFloatingPointVector<Vector2F<TScalar>, TScalar>,
+                           IVector2<Vector2F<TScalar>, TScalar>
+                           where TScalar : IFloatingPointIeee754<TScalar>
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2F<TScalar> UnitX { get; }
+    public static Vector2F<TScalar> UnitY { get; }
+
+    public static Vector2F<TScalar> Zero { get; }
+    public static Vector2F<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2F(TScalar x, TScalar y);
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, TScalar right);
+
+    public static Vector2F<TScalar> operator *(TScalar left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> value);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> value);
+
+    public TScalar LengthSquared { get; }
+    public TScalar Length();
+
+    public Vector2F<TScalar> Dot(Vector2F<TScalar> other);
+    public TScalar DistanceSquared(Vector2F<TScalar> other);
+    public TScalar Distance(Vector2F<TScalar> other);
+    public Vector2F<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
+
+    public static Vector2F<TScalar> Dot (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> DistanceSquared (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Distance (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Clamp (Vector2F<TScalar> value, Vector2F<TScalar> min, Vector2F<TScalar> max);
+    public static Vector2F<TScalar> Min (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Max (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Abs (Vector2F<TScalar> value);
+}
+```
+At the end of this proposal is the definition for the new interfaces used in the these classes.
```

Comment:
Missing the static utility classes.

Reply:
This feels more like an implementation detail for how some functions are given. I don't think we need it included in the proposal.

Reply:
I think this is specifically referring to `Vector2F<T>.Abs` vs `Vector2F.Abs<T>` - the latter the compiler can infer the T

Reply:
> This feels more like an implementation detail for how some functions are given. I don't think we need it included in the proposal.

The implementation for our primitive types will be explicit on the interface implementations and the actual code will be in the static classes. So this changes how you use the library unless you're implementing the interfaces yourself.

Reply:
Why would they be explicit? This feels like it’s unnecessarily limiting usability - if they want to call it on the type directly, why not? I _personally_ think that using the static class makes for cleaner code, but who am I to tell users how to write their code…

Reply:
You can't do `Vector2F<T>.Abs` in a generic context, you can only do `TVector.Abs(x)` or `x.Abs()` if you've defined a relevant extension method.

So having them be explicit makes it more consistent for non-generic code as well. You just always do `Vector2F.Abs(x)` and that can optionally be an extension, allowing `x.Abs()`.

It notably also helps with trimming and overall usability of the APIs, while making it more consistent for how people access it. -- I believe the static method would also block the extension method from working, if you didn't implement it explicitly.